### PR TITLE
[WFLY-19286] Add ServerSetupTasks to reload server to a desired stabi…

### PR DESCRIPTION
**Pull request by kabir**
…lity level

https://issues.redhat.com/browse/WFLY-19286

This needs the Arquillian upgrade in  https://github.com/wildfly/wildfly/pull/17832 merged before the tests can pass
____

More information about the [wildfly-bot[bot]](https://github.com/wildfly/wildfly-github-bot)

____

### CODE REVIEW

File: testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        // Do any additional setup from the sub-classes
+        doSetup(managementClient);
+    }
+
+    protected void doSetup(ManagementClient managementClient) throws Exception {
+
+    }
+
+    @Override
+    public void tearDown(ManagementClient managementClient, String containerId) throws Exception {
```
Comment (jamezp) [inline testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java:95]: Should this be a final method with a `doTearDown()` to ensure the snapshot is always restored?

**Comment by kabir** [comment]
@jamezp fixed (sorry for the delay)

File: testsuite/integration/smoke/src/test/java/org/jboss/as/test/smoke/stability/AbstractStabilityServerSetupTaskTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import org.jboss.as.test.integration.management.ManagementOperations;
+import org.jboss.as.version.Stability;
+import org.jboss.dmr.ModelNode;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.Assert;
+import org.junit.Test;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.CORE_SERVICE;
```
Comment (bstansberry) [inline testsuite/integration/smoke/src/test/java/org/jboss/as/test/smoke/stability/AbstractStabilityServerSetupTaskTest.java:22]: Static imports first.

File: testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+import java.util.Set;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.CORE_SERVICE;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.NAME;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.READ_OPERATION_DESCRIPTION_OPERATION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.READ_OPERATION_NAMES_OPERATION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.RELOAD_ENHANCED;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.STABILITY;
+import static org.jboss.as.server.controller.descriptions.ServerDescriptionConstants.SERVER_ENVIRONMENT;
+import static org.junit.Assert.fail;
```
Comment (bstansberry) [inline testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java:33]: Static imports first.

File: testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+import org.jboss.as.test.integration.management.ManagementOperations;
+import org.jboss.as.test.integration.management.util.ServerReload;
+import org.jboss.as.version.Stability;
+import org.jboss.dmr.ModelNode;
+import org.junit.Assert;
+import org.junit.Assume;
+
+import java.io.File;
+import java.util.HashSet;
+import java.util.Set;
```
Comment (bstansberry) [inline testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java:24]: java and javax after statics but before others

File: testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+import static org.jboss.as.server.controller.descriptions.ServerDescriptionConstants.SERVER_ENVIRONMENT;
+import static org.junit.Assert.fail;
+
+/**
+ * For tests that need to run under a specific server stability level,
+ * the server setup tasks from the inner classes can be used to change the stability level of the server to the desired level.
+ * Once the test is done, the original stability level is restored.
+ *
+ * In order to not pollute the configuration with XML from a different stability level following the run of the test,
+ * it takes a snapshot of the server configuration in the setup() method, and
```
Comment (bstansberry) [inline testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java:41]: and ..... ?

File: testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    public StabilityServerSetupSnapshotRestoreTasks(Stability desiredStability) {
+        this.desiredStability = desiredStability;
+    }
+
+    @Override
+    public final void setup(ManagementClient managementClient, String containerId) throws Exception {
+        // Make sure the desired stability level is one of the ones supported by the server
+        Set<Stability> supportedStabilityLevels = getSupportedStabilityLevels(managementClient);
+        Assume.assumeTrue(
+                String.format("%s is not a supported stability level", desiredStability, supportedStabilityLevels),
```
Comment (bstansberry) [inline testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java:60]: This is not a valid statement -- two vars but only one expression in the format string

File: testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    // Temporary duplication until we have the needed hooks in SnapshotServerSetupTasks to reload to different stability levels
+    private static AutoCloseable takeSnapshot(ManagementClient client, Stability reloadToStability) {
+        try {
+            ModelNode node = new ModelNode();
+            node.get(ModelDescriptionConstants.OP).set("take-snapshot");
+            ModelNode result = client.getControllerClient().execute(node);
+            if (!"success".equals(result.get(ClientConstants.OUTCOME).asString())) {
+                fail("Reload operation didn't finish successfully: " + result.asString());
```
Comment (bstansberry) [inline testsuite/shared/src/main/java/org/wildfly/test/stabilitylevel/StabilityServerSetupSnapshotRestoreTasks.java:201]: s/Reload/take-snapshot

**Review by bstansberry — CHANGES_REQUESTED** [review]
@kabir In general this LGTM to me but...

1) I have some minor comments
2) Please rebase as this branch now has two unrelated commits that have equivalents that are already in main.

Please feel free to merge this when ready; i.e. don't wait for my formal approval of the above as I'll not be around for a while.

**Pull request merged**